runtime/flang/fmtwrite.c

```diff
@@ -1750,13 +1750,114 @@ static bool call_format_double(int *result, int width, int format_char,
+                             int fraction_digits, int exponent_digits,
+                             int ESN_mode, int scale_factor,
+                             bool explicit_plus, bool comma_radix,
+                             bool elide_leading_spaces,
+                             bool elide_trailing_spaces, int rounding_mode,
+                             long double x)
+{
+  static int use_this_code_path = -1; /* unknown */
+  static int no_minus_zero = -1; /* unknown */
+
+  struct formatting_control control;
+
+  /* First call initializations */
+  if (use_this_code_path == -1)
+    use_this_code_path = __fortio_new_fp_formatter();
+  if (no_minus_zero == -1)
+    no_minus_zero = __fortio_no_minus_zero();
+
+  *result = 0;
+  if (!use_this_code_path)
+    return FALSE;
+  /* deal with format F0 */
+  if (format_char == 'F' && width == 0) {
+    control.format_F0 = 1;
+    width = G_REAL16_W + fraction_digits;
+  /* deal with format G0 and G0.d */
+  } else if (format_char == 'G' && width == 0) {
+    if (fraction_digits == 0) {
+      control.format_G0 = FORMAT_G0;
+      fraction_digits = G_REAL16_D;
+    } else
+      control.format_G0 = FORMAT_G0_D;
+    width = G_REAL16_W + fraction_digits;
+  } else {
+    control.format_G0 = control.format_F0 = 0;
+  }
+  control.rounding = rounding_mode;
+  control.format_char = format_char;
+  control.fraction_digits = fraction_digits;
+  control.exponent_digits = exponent_digits;
+  control.scale_factor = scale_factor; /* 1 for ES */
+  control.plus_sign = explicit_plus ? '+' : '\0';
+  control.point_char = comma_radix ? ',' : '.';
+  control.ESN_format = ESN_mode;
+  control.no_minus_zero = no_minus_zero;
+
+  if (elide_leading_spaces || elide_trailing_spaces || width > BUFFER_SIZE) {
+    /* Format into a buffer, chop spaces, and copy.  Eschew alloca(). */
+    char stack_buffer[BUFFER_SIZE];
+    char *emit = stack_buffer;
+    char *malloced_buffer = NULL;
+    char *pos = NULL;
+    memset(stack_buffer, ' ', BUFFER_SIZE);
+    if (width > sizeof stack_buffer &&
+        !(emit = malloced_buffer = malloc(((unsigned long)width)))) {
+      *result = __fortio_error(FIO_ENOMEM);
+    } else {
+      __fortio_format_quad(emit, width, &control, x);
+      if (elide_leading_spaces && (emit != NULL)) {
+        while (*emit == ' ' && width > 1) {
+          ++emit;
+          --width;
+        }
+      }
+      if (elide_trailing_spaces) {
+        width = 0;
+        pos = emit;
+        while(*pos != ' ' && *pos != '\0') {
+          ++pos;
+          ++width;
+        }
+      }
+      *result = fw_write_item(emit, width);
+      if (malloced_buffer != NULL)
+        free(malloced_buffer);
+    }
+  } else {
+    /* Format right into g->rec_buff, no copy */
+    char *emit = reserve_buffer(width);
+    if (emit == NULL) {
+      /* fw_check_size() failed, __fortio_error() was called */
+      *result = ERR_FLAG;
+    } else {
+      __fortio_format_quad(emit, width, &control, x);
+    }
```

Comment:
what happens to reserved memory pointed by the `emit` pointer?

Reply:
Nothing. Like the pre-existing `call_format_double`, `call_format_quad` simply requests space from the buffer `rec_buff` inside the global structure `gbl`, and calls `__fortio_format_quad` to write the formatted number to that space. The addition of quad-precision FP formatting has not changed how `rec_buff` is managed; the allocation and deallocation of the buffer is done somewhere else.